### PR TITLE
rgw: an empty tagset is allowed by S3

### PR DESCRIPTION
**Pull request by Rjerk**
Signed-off-by: Liu Lan <liulan_yewu@cmss.chinamobile.com>

https://docs.aws.amazon.com/AmazonS3/latest/dev/object-tagging.html

> A user might send an empty tag set in PutObjectTagging, which is allowed by this policy (an empty tag set in the request removes any existing tags on the object).

Test for this pr: https://github.com/ceph/s3-tests/pull/376

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by Rjerk** [comment]
jenkins retest this please

File: src/rgw/rgw_tag.cc [hunk 1]
```diff
@@ -35,6 +35,9 @@ int RGWObjTags::check_and_add_tag(const string&key, const string& val){
 }
 
 int RGWObjTags::set_from_string(const string& input){
+  if (input.empty()) {
+    return 0;
+  }
```
Comment (cbodley) [inline src/rgw/rgw_tag.cc:40]: wouldn't `boost::split()` just give us an empty vector, and we'd return success anyway?

File: src/rgw/rgw_tag_s3.cc [hunk 2]
```diff
@@ -30,7 +30,7 @@ void RGWObjTagEntry_S3::dump_xml(Formatter *f) const {
 void RGWObjTagSet_S3::decode_xml(XMLObj *obj) {
   vector<RGWObjTagEntry_S3> entries;
 
-  RGWXMLDecoder::decode_xml("Tag", entries, obj, true);
+  RGWXMLDecoder::decode_xml("Tag", entries, obj, false);
```
Comment (cbodley) [inline src/rgw/rgw_tag_s3.cc:33]: it's not clear what that last boolean flag means, can you say?

File: src/rgw/rgw_tag.cc (hunk 1 above)
Comment (Rjerk) [reply src/rgw/rgw_tag.cc:40]: no, it gives us a vector that contains an empty string.

File: src/rgw/rgw_tag_s3.cc (hunk 2 above)
Comment (Rjerk) [reply src/rgw/rgw_tag_s3.cc]: done

**Comment by Rjerk** [comment]
any pieces of advice?

File: src/rgw/rgw_tag.cc (hunk 1 above)
Comment (dang) [reply src/rgw/rgw_tag.cc:40]: In that case, maybe the simplest change would be to have check_and_add_tag() return success for empty key?

Comment (Rjerk) [reply src/rgw/rgw_tag.cc:40]: An empty tagset is well-formed, but an empty key is not.

**Review by dang — APPROVED** [review]

**Comment by dang** [comment]
https://pulpito.ceph.com/dang-2022-01-11_08:37:46-rgw-wip-dang-testing-distro-basic-smithi/

**Comment by votdev** [comment]
This seem not to work. Using `putObjectTagging` from the AWS SDK 2.1295.0 generates an URL like `http://127.0.0.1:4300/ddddd/file2.txt?tagging=` which the body `<Tagging xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><TagSet/></Tagging>` which fails with `<?xml version="1.0" encoding="UTF-8"?><Error><Code>MalformedXML</Code><BucketName>ddddd</BucketName><RequestId>tx00000fa3c64e6bd1b3296-0063c53afd-25a0b-default</RequestId><HostId>25a0b-default-default</HostId></Error>`.

**Pull request merged**